### PR TITLE
[WIP] Py3k (attempt 2)

### PR DESCRIPTION
**Pull request by kyleabeauchamp**
1.  Added futures to requirements.txt
2.  Ran `futurize  --nofix=division_safe -w stor/*.py stor/*/*.py`
3.  Global find and replace of `io.StringIO` with `io.BytesIO`

### CODE REVIEW

File: stor/base.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+from builtins import zip
+from builtins import map
```
Comment (lucaswiman) [inline stor/base.py]: These are pointless, since you're just iterating over these anyway.

File: requirements.txt [hunk 2]
```diff
@@ -2,3 +2,4 @@ boto3>=1.4.0
 cached_property
 python-keystoneclient>=1.8.1
 python-swiftclient
+futures>=3.0.5; python_version < '3.0'
```
Comment (lucaswiman) [inline requirements.txt]: The benefits of futures don't seem be matched by the cost of adding another dependency.

Comment (kyleabeauchamp) [inline requirements.txt]: It's already a dependency: https://github.com/openstack/python-swiftclient/blob/master/requirements.txt

File: stor/cli.py [hunk 3]
```diff
@@ -72,10 +72,14 @@
 Direct file transfer between OBS services or within one OBS service (server-side copy)
 is not yet supported.
 """
+from future import standard_library
+standard_library.install_aliases()
```
Comment (lucaswiman) [inline stor/cli.py]: Note that this introduces global monkey-patching, which is a fairly invasive thing for a dependency to do.

File: requirements.txt (hunk 2 above)
Comment (kyleabeauchamp) [inline requirements.txt]: I'm surprised they use *both* futures and six...

File: stor/cli.py [hunk 4]
```diff
@@ -72,10 +72,14 @@
...
+from builtins import str
+from builtins import range
```
Comment (lucaswiman) [inline stor/cli.py]: This is pointless since you're just iterating.

File: stor/cli.py [hunk 5]
```diff
@@ -72,10 +72,14 @@
 Direct file transfer between OBS services or within one OBS service (server-side copy)
 is not yet supported.
 """
+from future import standard_library
+standard_library.install_aliases()
+from builtins import str
```
Comment (lucaswiman) [inline stor/cli.py]: Note that most of the calls to `str` in this module are pointless (since they're `'%s' % str(something)`).

**Comment by jtratner** [comment]
Thanks for the PR - I'll take a look next week!

File: requirements.txt (hunk 2 above)
Comment (jtratner) [inline requirements.txt]: agreed, I think it's possible to do this without futures and rely on test suite to catch it.

File: stor/cli.py (hunk 3 above)
Comment (jtratner) [inline stor/cli.py]: agreed, let's see if we can do without it

File: stor/cli.py [hunk 6]
```diff
@@ -72,10 +72,14 @@
...
+from builtins import range
 
 import argparse
 import copy
-import ConfigParser
+import configparser
```
Comment (jtratner) [inline stor/cli.py]: this works? or is this from the aliases thing?

File: stor/cli.py (hunk 4 above)
Comment (jtratner) [inline stor/cli.py]: yeah I'd just get rid of this

File: stor/obs.py [hunk 7]
```diff
@@ -21,7 +25,7 @@ def wrapper(self, *args, **kwargs):
         func = getattr(self._buffer, attr_name)
         return func(*args, **kwargs)
     wrapper.__name__ = attr_name
-    wrapper.__doc__ = getattr(cStringIO.StringIO(), attr_name).__doc__
+    wrapper.__doc__ = getattr(io.BytesIO(), attr_name).__doc__
```
Comment (jtratner) [inline stor/obs.py]: potentially better to do:

```
try:
    from cStringIO import StringIO as BytesIO
except ImportError:
    from io import BytesIO
```

just to not change python 2.X behavior...

File: stor/settings.py [hunk 8]
```diff
@@ -1,6 +1,9 @@
+from future import standard_library
+standard_library.install_aliases()
+from builtins import object
```
Comment (jtratner) [inline stor/settings.py]: all these lines seem unnecessary

File: stor/tests/test_integration_s3.py [hunk 9]
```diff
@@ -1,4 +1,6 @@
-from cStringIO import StringIO
+from future import standard_library
+standard_library.install_aliases()
+from io import StringIO
```
Comment (jtratner) [inline stor/tests/test_integration_s3.py]: BytesIO?

File: stor/utils.py [hunk 10]
```diff
@@ -1,3 +1,6 @@
+from builtins import next
+from builtins import str
+from builtins import object
```
Comment (jtratner) [inline stor/utils.py]: these are prob all unnecessary, no?  or does it change 2.X behavior?

**Comment by kyleabeauchamp** [comment]
FWIW, it appears `_delegate_to_buffer` isn't quite happy with the naive `StringIO` to `BytesIO` conversion due to the lack of `next`.  

```
~/stor/stor/obs.py", line 28, in _delegate_to_buffer
    wrapper.__doc__ = getattr(io.BytesIO(), attr_name).__doc__
AttributeError: '_io.BytesIO' object has no attribute 'next'

```

**Comment by kyleabeauchamp** [comment]
NM, I think I found some comments in the docs about this.

**Comment by jtratner** [comment]
Easiest is to just add an if statement and not define "next" if python 3

**Comment by kyleabeauchamp** [comment]
Hmm, it does look like we have a lot of py2k-style `super(x,y)` calls scattered throughout.  We might want to strategize to deal with those in a uniform way?

**Comment by kyleabeauchamp** [comment]
In particular, our `super` chain eventually seems to lead to calling `object(pth)` on the root object type, which is forbidden in py3k.

**Comment by jtratner** [comment]
I bet it's because python 3 changed behavior when subclassing built in
types (in this case, str)

**Pull request closed without merging**